File: nta/engine/unittests/NetworkTest.cpp
/* ---------------------------------------------------------------------
 * Numenta Platform for Intelligent Computing (NuPIC)
 * Copyright (C) 2013, Numenta, Inc.  Unless you have an agreement
 * with Numenta, Inc., for a separate license for this software code, the
 * following terms and conditions apply:
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses.
 *
 * http://numenta.org/licenses/
 * ---------------------------------------------------------------------
 */

/** @file
 * Implementation of Network test
 */

#include "NetworkTest.hpp"

#include <nta/engine/Network.hpp>
#include <nta/engine/NuPIC.hpp>
#include <nta/engine/Region.hpp>
#include <nta/ntypes/Dimensions.hpp>

using namespace nta;


void NetworkTest::RunTests()
{
  test_nupic_auto_initialization();
  test_region_access();
  test_network_initialization();
  test_network_modification();
  test_phases();

}

void NetworkTest::test_nupic_auto_initialization()
{

  // Uninitialize NuPIC since this test checks auto-initialization
  // If shutdown fails, there is probably a problem with another test which 
  // is not cleaning up its networks. 
  if (NuPIC::isInitialized())
    NuPIC::shutdown();

  TEST(!NuPIC::isInitialized());
    
  // creating a network should auto-initialize NuPIC
  {
    Network net;
    TEST(NuPIC::isInitialized());
    Region *l1 = net.addRegion("level1", "TestNode", "");
    
    // Use l1 to avoid a compiler warning
    TESTEQUAL("level1", l1->getName());
    
    // Network still exists, so this should fail. 
    SHOULDFAIL(NuPIC::shutdown());
  }
  // net destructor has been called so we should be able to shut down NuPIC now
  NuPIC::shutdown();
}
  
void NetworkTest::test_region_access()
{
  Network net;
  SHOULDFAIL( net.addRegion("level1", "nonexistent_nodetype", "") );
  
  // Should be able to add a region 
  Region *l1 = net.addRegion("level1", "TestNode", "");

  TEST(l1->getNetwork() == &net);
  
  SHOULDFAIL(net.getRegions().getByName("nosuchregion"));
  
  // Make sure partial matches don't work
  SHOULDFAIL(net.getRegions().getByName("level"));
  
  Region* l1a = net.getRegions().getByName("level1");
  TEST(l1a == l1);
  
  // Should not be able to add a second region with the same name
  SHOULDFAIL(net.addRegion("level1", "TestNode", ""));

}


void NetworkTest::test_network_initialization()
{
  {
    Network net;
    net.initialize();
  }

  {
    Network net;
    Region *l1 = net.addRegion("level1", "TestNode", "");
    
    // Region does not yet have dimensions -- prevents network initialization
    SHOULDFAIL(net.initialize());
    SHOULDFAIL(net.run(1));
    
    Dimensions d;
    d.push_back(4);
    d.push_back(4);
    
    l1->setDimensions(d);
    
    // Should succeed since dimensions are now set
    net.initialize();
    net.run(1);

    Region *l2 = net.addRegion("level2", "TestNode", "");
    SHOULDFAIL(net.initialize());
    SHOULDFAIL(net.run(1));
    
    l2->setDimensions(d);
    net.run(1);

  }
  
}

void NetworkTest::test_network_modification()
{
  NTA_DEBUG << "Running network modification tests";
  {
    Network net;
    Region *l1 = net.addRegion("level1", "TestNode", "");

    // should have been added at phase0
    std::set<UInt32> phases = net.getPhases("level1");
    TESTEQUAL((UInt32)1, phases.size());
    TEST(phases.find(0) != phases.end());

    Dimensions d;
    d.push_back(4);
    d.push_back(4);
    l1->setDimensions(d);

    net.addRegion("level2", "TestNode", "");

    // should have been added at phase1
    phases = net.getPhases("level2");
    TESTEQUAL((UInt32)1, phases.size());
    TEST(phases.find(1) != phases.end());


    net.link("level1", "level2", "TestFanIn2", "");
    
    const Collection<Region*>& regions = net.getRegions();

    TESTEQUAL((UInt32)2, regions.getCount());

    // Should succeed since dimensions are now set
    net.initialize();
    net.run(1);
    Region* l2 = regions.getByName("level2");
    Dimensions d2 = l2->getDimensions();
    TESTEQUAL((UInt32)2, d2.size());
    TESTEQUAL((UInt32)2, d2[0]);
    TESTEQUAL((UInt32)2, d2[1]);
    
    SHOULDFAIL(net.removeRegion("doesntexist"));

    net.removeRegion("level2");
    // net now only contains level1
    TESTEQUAL((UInt32)1, regions.getCount());
    SHOULDFAIL(regions.getByName("level2"));
    
    // network requires initialization, but this will auto-initialize
    net.run(1);

    TEST(l1 == regions.getByName("level1"));
    l2 = net.addRegion("level2", "TestNode", "");

    // should have been added at phase1
    phases = net.getPhases("level2");
    TESTEQUAL((UInt32)1, phases.size());
    TEST(phases.find(1) != phases.end());

    // network requires initialization, but can't be initialized
    // because level2 is not initialized
    SHOULDFAIL(net.run(1));

    net.link("level1", "level2", "TestFanIn2", "");
    
    // network can be initialized now
    net.run(1);

    TESTEQUAL((UInt32)2, regions.getCount());
    TEST(l2 == regions.getByName("level2"));

    d2 = l2->getDimensions();
    TESTEQUAL((UInt32)2, d2.size());
    TESTEQUAL((UInt32)2, d2[0]);
    TESTEQUAL((UInt32)2, d2[1]);
               
    // add a third region
    Region* l3 = net.addRegion("level3", "TestNode", "");

    // should have been added at phase 2
    phases = net.getPhases("level3");
    TESTEQUAL((UInt32)1, phases.size());
    TEST(phases.find(2) != phases.end());

    TESTEQUAL((UInt32)3, regions.getCount());

    // network requires initialization, but can't be initialized
    // because level3 is not initialized
    SHOULDFAIL(net.run(1));

    net.link("level2", "level3", "TestFanIn2", "");
    net.initialize();
    d2 = l3->getDimensions();
    TESTEQUAL((UInt32)2, d2.size());
    TESTEQUAL((UInt32)1, d2[0]);
    TESTEQUAL((UInt32)1, d2[1]);
    
    // try to remove a region whose outputs are connected
    // this should fail because it would leave the network
    // unrunnable
    SHOULDFAIL(net.removeRegion("level2"));
    TESTEQUAL((UInt32)3, regions.getCount());
    SHOULDFAIL(net.removeRegion("level1"));
    TESTEQUAL((UInt32)3, regions.getCount());

    // this should be ok
    net.removeRegion("level3");
    TESTEQUAL((UInt32)2, regions.getCount());

    net.removeRegion("level2");
    net.removeRegion("level1");
    TESTEQUAL((UInt32)0, regions.getCount());
    
    // build up the network again -- slightly differently with 
    // l1->l2 and l1->l3
    l1 = net.addRegion("level1", "TestNode", "");
    l1->setDimensions(d); 
    net.addRegion("level2", "TestNode", "");
    net.addRegion("level3", "TestNode", "");
    net.link("level1", "level2", "TestFanIn2", "");
    net.link("level1", "level3", "TestFanIn2", "");
    net.initialize();

    // build it up one more time and let the destructor take care of it
    net.removeRegion("level2");
    net.removeRegion("level3");
    net.run(1);

    l2 = net.addRegion("level2", "TestNode", "");
    l3 = net.addRegion("level3", "TestNode", "");
    // try links in reverse order
    net.link("level2", "level3", "TestFanIn2", "");
    net.link("level1", "level2", "TestFanIn2", "");
    net.initialize();
    d2 = l3->getDimensions();
    TESTEQUAL((UInt32)2, d2.size());
    TESTEQUAL((UInt32)1, d2[0]);
    TESTEQUAL((UInt32)1, d2[1]);

    d2 = l2->getDimensions();
    TESTEQUAL((UInt32)2, d2.size());
    TESTEQUAL((UInt32)2, d2[0]);
    TESTEQUAL((UInt32)2, d2[1]);
    
    // now let the destructor remove everything
    
  }


  { 
    // unlinking tests
    NTA_DEBUG << "Running unlinking tests";
    Network net;
    net.addRegion("level1", "TestNode", "");
    net.addRegion("level2", "TestNode", "");
    Dimensions d;
    d.push_back(4);
    d.push_back(2);
    net.getRegions().getByName("level1")->setDimensions(d);

    net.link("level1", "level2", "TestFanIn2", "");
    TEST(net.getRegions().getByName("level2")->getDimensions().isUnspecified());

    SHOULDFAIL(net.removeLink("level1", "level2", "outputdoesnotexist", "bottomUpIn"));
    SHOULDFAIL(net.removeLink("level1", "level2", "bottomUpOut", "inputdoesnotexist"));
    SHOULDFAIL(net.removeLink("level1", "leveldoesnotexist"));
    SHOULDFAIL(net.removeLink("leveldoesnotexist", "level2"));

    // remove the link from the uninitialized network
    net.removeLink("level1", "level2");
    TEST(net.getRegions().getByName("level2")->getDimensions().isUnspecified());

    SHOULDFAIL(net.removeLink("level1", "level2"));

    // remove, specifying output/input names
    net.link("level1", "level2", "TestFanIn2", "");
    net.removeLink("level1", "level2", "bottomUpOut", "bottomUpIn");
    SHOULDFAIL(net.removeLink("level1", "level2", "bottomUpOut", "bottomUpIn"));

    net.link("level1", "level2", "TestFanIn2", "");
    net.removeLink("level1", "level2", "bottomUpOut");
    SHOULDFAIL(net.removeLink("level1", "level2", "bottomUpOut"))
    
    // add the link back and initialize (inducing dimensions)
    net.link("level1", "level2", "TestFanIn2", "");
    net.initialize();

    d = net.getRegions().getByName("level2")->getDimensions();
    TESTEQUAL((UInt32)2, d.size());
    TESTEQUAL((UInt32)2, d[0]);
    TESTEQUAL((UInt32)1, d[1]);
    
    // remove the link. This will fail because we can't 
    // remove a link to an initialized region
    SHOULDFAIL_WITH_MESSAGE(net.removeLink("level1", "level2"), 
                            "Cannot remove link [level1.bottomUpOut (region dims: [4 2])  to level2.bottomUpIn (region dims: [2 1])  type: TestFanIn2] because destination region level2 is initialized. Remove the region first.");
    
  }

}

typedef std::vector<std::string> callbackData;
callbackData mydata;

void testCallback(Network* net, UInt64 iteration, void* data)
{
  callbackData& thedata = *(static_cast<callbackData*>(data));
  // push region names onto callback data
  const nta::Collection<Region*>& regions = net->getRegions();
  for (size_t i = 0; i < regions.getCount(); i++)
  {
    thedata.push_back(regions.getByIndex(i).first);
  }
}


std::vector<std::string> computeHistory;
static void recordCompute(const std::string& name)
{
  computeHistory.push_back(name);
}


void NetworkTest::test_phases()
{
  {
    Network net;
  
    // should auto-initialize with max phase
    Region *l1 = net.addRegion("level1", "TestNode", "");
    // Use l1 to avoid a compiler warning
    TESTEQUAL("level1", l1->getName());

    std::set<UInt32> phaseSet = net.getPhases("level1");
    TESTEQUAL((UInt32)1, phaseSet.size());
    TEST(phaseSet.find(0) != phaseSet.end());


    Region *l2 = net.addRegion("level2", "TestNode", "");
    TESTEQUAL("level2", l2->getName());
    phaseSet = net.getPhases("level2");
    TEST(phaseSet.size() == 1);
    TEST(phaseSet.find(1) != phaseSet.end());

    SHOULDFAIL(net.initialize());
  
    Dimensions d;
    d.push_back(2);
    d.push_back(2);

    l1->setDimensions(d);
    l2->setDimensions(d);
    net.initialize();
    l1->setParameterUInt64("computeCallback", (UInt64)recordCompute);
    l2->setParameterUInt64("computeCallback", (UInt64)recordCompute);

    computeHistory.clear();
    net.run(2);
    TESTEQUAL((UInt32)4, computeHistory.size());
    // use at() to throw an exception if out of range
    TESTEQUAL("level1", computeHistory.at(0));
    TESTEQUAL("level2", computeHistory.at(1));
    TESTEQUAL("level1", computeHistory.at(2));
    TESTEQUAL("level2", computeHistory.at(3));
    computeHistory.clear();

    phaseSet.clear();
    phaseSet.insert(0);
    phaseSet.insert(2);
    net.setPhases("level1", phaseSet);
    net.run(2);
    TESTEQUAL((UInt32)6, computeHistory.size());
    if (computeHistory.size() == 6)
    {
      TESTEQUAL("level1", computeHistory.at(0));
      TESTEQUAL("level2", computeHistory.at(1));
      TESTEQUAL("level1", computeHistory.at(2));
      TESTEQUAL("level1", computeHistory.at(3));
      TESTEQUAL("level2", computeHistory.at(4));
      TESTEQUAL("level1", computeHistory.at(5));
    }
    computeHistory.clear();
  }
  {
    // tests for min/max phase
    Network n;
    UInt32 minPhase = n.getMinPhase();
    UInt32 maxPhase = n.getMaxPhase();

    TESTEQUAL((UInt32)0, minPhase);
    TESTEQUAL((UInt32)0, maxPhase);
    
    SHOULDFAIL(n.setMinEnabledPhase(1));
    SHOULDFAIL(n.setMaxEnabledPhase(1));
    Region *l1 = n.addRegion("level1", "TestNode", "");
    Region *l2 = n.addRegion("level2", "TestNode", "");
    Region *l3 = n.addRegion("level3", "TestNode", "");
    Dimensions d;
    d.push_back(1);
    l1->setDimensions(d);
    l2->setDimensions(d);
    l3->setDimensions(d);

    n.initialize();

    l1->setParameterUInt64("computeCallback", (UInt64)recordCompute);
    l2->setParameterUInt64("computeCallback", (UInt64)recordCompute);
    l3->setParameterUInt64("computeCallback", (UInt64)recordCompute);

    minPhase = n.getMinEnabledPhase();
    maxPhase = n.getMaxEnabledPhase();

    TESTEQUAL((UInt32)0, minPhase);
    TESTEQUAL((UInt32)2, maxPhase);

    computeHistory.clear();
    n.run(2);
    TESTEQUAL((UInt32)6, computeHistory.size());
    TESTEQUAL("level1", computeHistory.at(0));
    TESTEQUAL("level2", computeHistory.at(1));
    TESTEQUAL("level3", computeHistory.at(2));
    TESTEQUAL("level1", computeHistory.at(3));
    TESTEQUAL("level2", computeHistory.at(4));
    TESTEQUAL("level3", computeHistory.at(5));
    
    
    n.setMinEnabledPhase(0);
    n.setMaxEnabledPhase(1);
    computeHistory.clear();
    n.run(2);
    TESTEQUAL((UInt32)4, computeHistory.size());
    TESTEQUAL("level1", computeHistory.at(0));
    TESTEQUAL("level2", computeHistory.at(1));
    TESTEQUAL("level1", computeHistory.at(2));
    TESTEQUAL("level2", computeHistory.at(3));

    n.setMinEnabledPhase(1);
    n.setMaxEnabledPhase(1);
    computeHistory.clear();
    n.run(2);
    TESTEQUAL((UInt32)2, computeHistory.size());
    TESTEQUAL("level2", computeHistory.at(0));
    TESTEQUAL("level2", computeHistory.at(1));

    // reset to full network
    n.setMinEnabledPhase(0);
    n.setMaxEnabledPhase(n.getMaxPhase());
    computeHistory.clear();
    n.run(2);
    TESTEQUAL((UInt32)6, computeHistory.size());
    if (computeHistory.size() == 6)
    {
      TESTEQUAL("level1", computeHistory.at(0));
      TESTEQUAL("level2", computeHistory.at(1));
      TESTEQUAL("level3", computeHistory.at(2));
      TESTEQUAL("level1", computeHistory.at(3));
      TESTEQUAL("level2", computeHistory.at(4));
      TESTEQUAL("level3", computeHistory.at(5));
    }
    // max < min; allowed, but network should not run
    n.setMinEnabledPhase(1);
    n.setMaxEnabledPhase(0);
    computeHistory.clear();
    n.run(2);
    TESTEQUAL((UInt32)0, computeHistory.size());

    // max > network max
    SHOULDFAIL(n.setMaxEnabledPhase(4));
    
    std::set<UInt32> phases;
    phases.insert(4);
    phases.insert(6);
    n.setPhases("level2", phases);
    n.removeRegion("level1");
    // we now have: level2: 4, 6  level3: 2
    
    minPhase = n.getMinPhase();
    maxPhase = n.getMaxPhase();

    TESTEQUAL((UInt32)2, minPhase);
    TESTEQUAL((UInt32)6, maxPhase);

    computeHistory.clear();
    n.run(2);

    TESTEQUAL((UInt32)6, computeHistory.size());
    TESTEQUAL("level3", computeHistory.at(0));
    TESTEQUAL("level2", computeHistory.at(1));
    TESTEQUAL("level2", computeHistory.at(2));
    TESTEQUAL("level3", computeHistory.at(3));
    TESTEQUAL("level2", computeHistory.at(4));
    TESTEQUAL("level2", computeHistory.at(5));
    
    
  }

  {
    // callback test
    Network n;
    n.addRegion("level1", "TestNode", "");
    n.addRegion("level2", "TestNode", "");
    n.addRegion("level3", "TestNode", "");
    Dimensions d;
    d.push_back(1);
    n.getRegions().getByName("level1")->setDimensions(d);
    n.getRegions().getByName("level2")->setDimensions(d);
    n.getRegions().getByName("level3")->setDimensions(d);


    Collection<Network::callbackItem>& callbacks = n.getCallbacks();
    Network::callbackItem callback(testCallback, (void*)(&mydata));
    callbacks.add("Test Callback", callback);

    n.run(2);
    TESTEQUAL((UInt32)6, mydata.size());
    TESTEQUAL("level1", mydata[0]);
    TESTEQUAL("level2", mydata[1]);
    TESTEQUAL("level3", mydata[2]);
    TESTEQUAL("level1", mydata[3]);
    TESTEQUAL("level2", mydata[4]);
    TESTEQUAL("level3", mydata[5]);

  }    
}








  

File: tests/unit/nupic/engine/network_test.py
# ----------------------------------------------------------------------
# Numenta Platform for Intelligent Computing (NuPIC)
# Copyright (C) 2014-2017, Numenta, Inc.  Unless you have an agreement
# with Numenta, Inc., for a separate license for this software code, the
# following terms and conditions apply:
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero Public License version 3 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Affero Public License for more details.
#
# You should have received a copy of the GNU Affero Public License
# along with this program.  If not, see http://www.gnu.org/licenses.
#
# http://numenta.org/licenses/
# ----------------------------------------------------------------------

import sys
from mock import patch
import unittest2 as unittest

from nupic import engine
from nupic.bindings.regions.TestNode import TestNode
from nupic.regions.sp_region import SPRegion



class NetworkTest(unittest.TestCase):


  @unittest.skipIf(sys.platform.lower().startswith("win"),
                   "Not supported on Windows, yet!")
  def testErrorHandling(self):
    n = engine.Network()

    # Test trying to add non-existent node
    with self.assertRaises(Exception) as cm:
      n.addRegion('r', 'py.NonExistingNode', '')

    self.assertEqual(cm.exception.message, "Matching Python module for NonExistingNode not found.")

    orig_import = __import__
    def import_mock(name, *args):
      if name == "nupic.regions.unimportable_node":
        raise SyntaxError("invalid syntax (unimportable_node.py, line 5)")

      return orig_import(name, *args)

    with patch('__builtin__.__import__', side_effect=import_mock):
      # Test failure during import
      with self.assertRaises(Exception) as cm:
        n.addRegion('r', 'py.UnimportableNode', '')

      self.assertEqual(cm.exception.message, "invalid syntax (unimportable_node.py, line 5)")

    # Test failure in the __init__() method
    with self.assertRaises(Exception) as cm:
      n.addRegion('r', 'py.TestNode', '{ failInInit: 1 }')

    self.assertEqual(cm.exception.message, "TestNode.__init__() Failing on purpose as requested")

    # Test failure inside the compute() method
    with self.assertRaises(Exception) as cm:
      r = n.addRegion('r', 'py.TestNode', '{ failInCompute: 1 }')
      r.dimensions = engine.Dimensions([4, 4])
      n.initialize()
      n.run(1)

    self.assertEqual(str(cm.exception),
      'TestNode.compute() Failing on purpose as requested')

    # Test failure in the static getSpec
    from nupic.bindings.regions.TestNode import TestNode
    TestNode._failIngetSpec = True

    with self.assertRaises(Exception) as cm:
      TestNode.getSpec()

    self.assertEqual(str(cm.exception),
      'Failing in TestNode.getSpec() as requested')

    del TestNode._failIngetSpec


  def testGetSpecFromType(self):
    ns = engine.Region.getSpecFromType('py.SPRegion')
    p = ns.parameters['breakPdb']
    self.assertEqual(p.accessMode, 'ReadWrite')


  def testOneRegionNetwork(self):
    n = engine.Network()

    print "Number of regions in new network: %d" % len(n.regions)
    self.assertEqual(len(n.regions), 0)

    print "Adding level1SP"
    level1SP = n.addRegion("level1SP", "TestNode", "")
    print "Current dimensions are: %s" % level1SP.dimensions
    print "Number of regions in network: %d" % len(n.regions)

    self.assertEqual(len(n.regions), 1)
    self.assertEqual(len(n.regions), len(n.regions))

    print 'Node type: ', level1SP.type

    print("Attempting to initialize net when "
           "one region has unspecified dimensions")
    print "Current dimensions are: %s" % level1SP.dimensions

    with self.assertRaises(Exception):
      n.initialize()

    # Test Dimensions
    level1SP.dimensions = engine.Dimensions([4, 4])
    print "Set dimensions of level1SP to %s" % str(level1SP.dimensions)

    n.initialize()

    # Test Array
    a = engine.Array('Int32', 10)
    self.assertEqual(a.getType(), 'Int32')
    self.assertEqual(len(a), 10)
    import nupic
    self.assertEqual(type(a), nupic.bindings.engine_internal.Int32Array)

    for i in range(len(a)):
      a[i] = i

    for i in range(len(a)):
      self.assertEqual(type(a[i]), int)
      self.assertEqual(a[i], i)
      print i,
    print

    # --- Test Numpy Array
    print 'Testing Numpy Array'
    a = engine.Array('Byte', 15)
    print len(a)
    for i in range(len(a)):
      a[i] = ord('A') + i

    for i in range(len(a)):
      print a[i], ord('A') + i
      self.assertEqual(ord(a[i]), ord('A') + i)
    print

    print 'before asNumpyarray()'
    na = a.asNumpyArray()
    print 'after asNumpyarray()'

    self.assertEqual(na.shape, (15,))
    print 'na.shape:', na.shape
    na = na.reshape(5, 3)
    self.assertEqual(na.shape, (5, 3))
    print 'na.shape:', na.shape
    for i in range(5):
      for j in range(3):
        print chr(na[i, j]), ' ',
      print
    print


    # --- Test get/setParameter for Int64 and Real64
    print '---'
    print 'Testing get/setParameter for Int64/Real64'
    val = level1SP.getParameterInt64('int64Param')
    rval = level1SP.getParameterReal64('real64Param')
    print 'level1SP.int64Param = ', val
    print 'level1SP.real64Param = ', rval

    val = 20
    level1SP.setParameterInt64('int64Param', val)
    val = 0
    val = level1SP.getParameterInt64('int64Param')
    print 'level1SP.int64Param = ', val, ' after setting to 20'

    rval = 30.1
    level1SP.setParameterReal64('real64Param', rval)
    rval = 0.0
    rval = level1SP.getParameterReal64('real64Param')
    print 'level1SP.real64Param = ', rval, ' after setting to 30.1'

    # --- Test array parameter
    # Array a will be allocated inside getParameter
    print '---'
    print 'Testing get/setParameterArray'
    a = engine.Array('Int64', 4)
    level1SP.getParameterArray("int64ArrayParam", a)
    print 'level1SP.int64ArrayParam size = ', len(a)
    print 'level1SP.int64ArrayParam = [ ',
    for i in range(len(a)):
      print a[i],

    print ']'
    #
    # --- test setParameter of an Int64 Array ---
    print 'Setting level1SP.int64ArrayParam to [ 1 2 3 4 ]'
    a2 = engine.Array('Int64', 4)
    for i in range(4):
      a2[i] = i + 1

    level1SP.setParameterArray('int64ArrayParam', a2)

    # get the value of int64ArrayParam after the setParameter call.
    # The array a owns its buffer, so we can call releaseBuffer if we
    # want, but the buffer should be reused if we just pass it again.
    #// a.releaseBuffer();
    level1SP.getParameterArray('int64ArrayParam', a)
    print 'level1SP.int64ArrayParam size = ', len(a)
    print 'level1SP.int64ArrayParam = [ ',
    for i in range(len(a)):
      print a[i],
    print ']'

    level1SP.compute()

    print "Running for 2 iteraitons"
    n.run(2)


    # --- Test input/output access
    #
    # Getting access via zero-copy
    with self.assertRaises(Exception):
      level1SP.getOutputData('doesnotexist')

    output = level1SP.getOutputData('bottomUpOut')
    print 'Element count in bottomUpOut is ', len(output)
    # set the actual output
    output[11] = 7777
    output[12] = 54321


    # Create a reshaped view of the numpy array
    # original output is 32x1 -- 16 nodes, 2 elements per node
    # Reshape to 8 rows, 4 columns
    numpy_output2 = output.reshape(8, 4)

    # Make sure the original output, the numpy array and the reshaped numpy view
    # are all in sync and access the same underlying memory.
    numpy_output2[1, 0] = 5555
    self.assertEqual(output[4], 5555)

    output[5] = 3333
    self.assertEqual(numpy_output2[1, 1], 3333)
    numpy_output2[1, 2] = 4444

    # --- Test doc strings
    # TODO: commented out because I'm not sure what to do with these
    # now that regions have been converted to the Collection class.
    # print
    # print "Here are some docstrings for properties and methods:"
    # for name in ('regionCount', 'getRegionCount', 'getRegionByName'):
    #   x = getattr(engine.Network, name)
    #   if isinstance(x, property):
    #     print 'property Network.{0}: "{1}"'.format(name, x.__doc__)
    #   else:
    #     print 'method Network.{0}(): "{1}"'.format(name, x.__doc__)

    # Typed methods should return correct type
    print "real64Param: %.2f" % level1SP.getParameterReal64("real64Param")

    # Uncomment to get performance for getParameter

    if 0:
      import time
      t1 = time.time()
      t1 = time.time()
      for i in xrange(0, 1000000):
        # x = level1SP.getParameterInt64("int64Param")   # buffered
        x = level1SP.getParameterReal64("real64Param")   # unbuffered
      t2 = time.time()

      print "Time for 1M getParameter calls: %.2f seconds" % (t2 - t1)


  def testTwoRegionNetwork(self):
    n = engine.Network()

    region1 = n.addRegion("region1", "TestNode", "")
    region2 = n.addRegion("region2", "TestNode", "")

    names = [region[0] for region in n.regions]
    self.assertEqual(names, ['region1', 'region2'])
    print n.getPhases('region1')
    self.assertEqual(n.getPhases('region1'), (0,))
    self.assertEqual(n.getPhases('region2'), (1,))

    n.link("region1", "region2", "TestFanIn2", "")

    print "Initialize should fail..."
    with self.assertRaises(Exception):
      n.initialize()

    print "Setting region1 dims"
    r1dims = engine.Dimensions([6, 4])
    region1.setDimensions(r1dims)

    print "Initialize should now succeed"
    n.initialize()

    r2dims = region2.dimensions
    self.assertEqual(len(r2dims), 2)
    self.assertEqual(r2dims[0], 3)
    self.assertEqual(r2dims[1], 2)

    # Negative test
    with self.assertRaises(Exception):
      region2.setDimensions(r1dims)


  def testDelayedLink(self):
    n = engine.Network()

    region1 = n.addRegion("region1", "TestNode", "")
    region2 = n.addRegion("region2", "TestNode", "")

    names = []

    propagationDelay = 2
    n.link("region1", "region2", "TestFanIn2", "",
           propagationDelay=propagationDelay)

    r1dims = engine.Dimensions([6, 4])
    region1.setDimensions(r1dims)

    n.initialize()

    outputArrays = []
    inputArrays = []

    iterations = propagationDelay + 2
    for i in xrange(iterations):
      n.run(1)

      if i < iterations - propagationDelay:
        outputArrays.append(list(region1.getOutputData("bottomUpOut")))

      if i < propagationDelay:
        # Pre-initialized delay elements should be arrays of all 0's
        outputArrays.insert(i, [0.0] * len(outputArrays[0]))

      inputArrays.append(list(region2.getInputData("bottomUpIn")))

    self.assertListEqual(inputArrays, outputArrays)


  def testInputsAndOutputs(self):
    n = engine.Network()

    region1 = n.addRegion("region1", "TestNode", "")
    region2 = n.addRegion("region2", "TestNode", "")
    region1.setDimensions(engine.Dimensions([6, 4]))
    n.link("region1", "region2", "TestFanIn2", "")
    n.initialize()

    r1_output = region1.getOutputData("bottomUpOut")

    region1.compute()
    print "Region 1 output after first iteration:"
    print "r1_output:", r1_output

    region2.prepareInputs()
    r2_input = region2.getInputData("bottomUpIn")
    print "Region 2 input after first iteration:"
    print 'r2_input:', r2_input


  def testNodeSpec(self):
    n = engine.Network()
    r = n.addRegion("region", "TestNode", "")

    print r.getSpec()


  @unittest.skipIf(sys.platform.lower().startswith("win"),
                   "Not supported on Windows, yet!")
  def testPyNodeGetSetParameter(self):
    n = engine.Network()

    r = n.addRegion("region", "py.TestNode", "")

    print "Setting region1 dims"
    r.dimensions = engine.Dimensions([6, 4])

    print "Initialize should now succeed"
    n.initialize()

    result = r.getParameterReal64('real64Param')
    self.assertEqual(result, 64.1)

    r.setParameterReal64('real64Param', 77.7)

    result = r.getParameterReal64('real64Param')
    self.assertEqual(result, 77.7)


  @unittest.skipIf(sys.platform.lower().startswith("win"),
                   "Not supported on Windows, yet!")
  def testPyNodeGetNodeSpec(self):
    n = engine.Network()

    r = n.addRegion("region", "py.TestNode", "")

    print "Setting region1 dims"
    r.setDimensions(engine.Dimensions([6, 4]))

    print "Initialize should now succeed"
    n.initialize()

    ns = r.spec

    self.assertEqual(len(ns.inputs), 1)
    i = ns.inputs['bottomUpIn']
    self.assertEqual(i.description, 'Primary input for the node')

    self.assertEqual(len(ns.outputs), 1)
    i = ns.outputs['bottomUpOut']
    self.assertEqual(i.description, 'Primary output for the node')


  @unittest.skipIf(sys.platform.lower().startswith("win"),
                   "Not supported on Windows, yet!")
  def testTwoRegionPyNodeNetwork(self):
    n = engine.Network()

    region1 = n.addRegion("region1", "py.TestNode", "")
    region2 = n.addRegion("region2", "py.TestNode", "")

    n.link("region1", "region2", "TestFanIn2", "")

    print "Initialize should fail..."
    with self.assertRaises(Exception):
      n.initialize()

    print "Setting region1 dims"
    r1dims = engine.Dimensions([6, 4])
    region1.setDimensions(r1dims)

    print "Initialize should now succeed"
    n.initialize()

    r2dims = region2.dimensions
    self.assertEqual(len(r2dims), 2)
    self.assertEqual(r2dims[0], 3)
    self.assertEqual(r2dims[1], 2)


  def testGetRegion(self):
    n = engine.Network()
    n.addRegion("region1", "py.TestNode", "")

    region = n.getRegionsByType(TestNode)[0]
    self.assertEqual(type(region.getSelf()), TestNode)

    self.assertEqual(n.getRegionsByType(SPRegion), [])
